File: ios/VisionCameraCropperFactory.swift
import AVFoundation
import CoreImage
import Foundation
import NitroModules
import UIKit
import VisionCamera

class VisionCameraCropperFactory: HybridVisionCameraCropperSpec {

  func rotateImage(base64: String, degree: Double) throws -> String {
    guard let image = Self.decodeBase64(base64) else {
      throw RuntimeError.error(withMessage: "Invalid base64")
    }
    let rotated = degree != 0 ? Self.rotate(image: image, degree: CGFloat(degree)) : image
    guard let result = Self.encodeBase64(rotated) else {
      throw RuntimeError.error(withMessage: "Failed to encode image")
    }
    return result
  }

  func cropFrame(frame: any HybridFrameSpec, options: CropConfig) throws -> CropResult {
    guard let nativeFrame = frame as? any NativeFrame else {
      throw RuntimeError.error(withMessage: "Frame is not a NativeFrame!")
    }
    guard let sampleBuffer = nativeFrame.sampleBuffer else {
      throw RuntimeError.error(withMessage: "Frame has already been disposed!")
    }

    let uiOrientation = Self.uiImageOrientation(
      from: frame.orientation,
      isMirrored: frame.isMirrored
    )
    var image = try Self.uiImage(from: sampleBuffer, orientation: uiOrientation)

    // FOR SOME REASON NEED TO ROTATE WHEN ORIENTATION IS LEFT
    if frame.orientation == .left {
      image = Self.rotate(image: image, degree: 180)
    }

    if let cropRegion = options.cropRegion {
      image = try Self.crop(image: image, region: cropRegion)
    }

    var base64: String?
    if options.includeImageBase64 == true {
      base64 = Self.encodeBase64(image)
    }

    var path: String?
    if options.saveAsFile == true {
      path = try Self.saveToTempFile(image)
    }

    return CropResult(base64: base64, path: path)
  }

  // MARK: - Helpers

  private static let ciContext = CIContext(options: [.useSoftwareRenderer: false])

  private static func uiImageOrientation(
    from orientation: CameraOrientation,
    isMirrored: Bool
  ) -> UIImage.Orientation {
    switch orientation {
    case .up:
      return isMirrored ? .upMirrored : .up
    case .down:
      return isMirrored ? .downMirrored : .down
    case .left:
      return isMirrored ? .leftMirrored : .left
    case .right:
      return isMirrored ? .rightMirrored : .right
    }
  }

  private static func uiImage(
    from sampleBuffer: CMSampleBuffer,
    orientation: UIImage.Orientation
  ) throws -> UIImage {
    guard let imageBuffer = sampleBuffer.imageBuffer else {
      throw RuntimeError.error(withMessage: "This Frame does not have a PixelBuffer!")
    }
    let attachments = CMCopyDictionaryOfAttachments(
      allocator: kCFAllocatorDefault,
      target: sampleBuffer,
      attachmentMode: kCMAttachmentMode_ShouldPropagate
    )
    let ciAttachments = attachments as? [CIImageOption: Any]
    let ciImage = CIImage(cvPixelBuffer: imageBuffer, options: ciAttachments)
    guard let cgImage = ciContext.createCGImage(ciImage, from: ciImage.extent) else {
      throw RuntimeError.error(withMessage: "Failed to copy CIImage into CGImage!")
    }
    return UIImage(cgImage: cgImage, scale: 1, orientation: orientation)
  }

  private static func decodeBase64(_ base64: String) -> UIImage? {
    guard let data = Data(base64Encoded: base64, options: .ignoreUnknownCharacters) else {
      return nil
    }
    return UIImage(data: data)
  }

  private static func encodeBase64(_ image: UIImage) -> String? {
    image.jpegData(compressionQuality: 1.0)?.base64EncodedString()
  }

  private static func rotate(image: UIImage, degree: CGFloat) -> UIImage {
    let radians = degree * (.pi / 180.0)
    let rotatedSize = CGRect(origin: .zero, size: image.size)
      .applying(CGAffineTransform(rotationAngle: radians))
      .integral.size
    UIGraphicsBeginImageContext(rotatedSize)
    defer { UIGraphicsEndImageContext() }
    guard let context = UIGraphicsGetCurrentContext() else {
      return image
    }
    let origin = CGPoint(x: rotatedSize.width / 2.0, y: rotatedSize.height / 2.0)
    context.translateBy(x: origin.x, y: origin.y)
    context.rotate(by: radians)
    image.draw(
      in: CGRect(
        x: -image.size.width / 2.0,
        y: -image.size.height / 2.0,
        width: image.size.width,
        height: image.size.height
      )
    )
    return UIGraphicsGetImageFromCurrentImageContext() ?? image
  }

  private static func crop(image: UIImage, region: CropRegion) throws -> UIImage {
    guard let cgImage = image.cgImage else {
      throw RuntimeError.error(withMessage: "Failed to get CGImage!")
    }
    let imgWidth = Double(cgImage.width)
    let imgHeight = Double(cgImage.height)
    let left = region.left / 100.0 * imgWidth
    let top = region.top / 100.0 * imgHeight
    let width = region.width / 100.0 * imgWidth
    let height = region.height / 100.0 * imgHeight
    let cropRect = CGRect(x: left, y: top, width: width, height: height).integral
    guard let cropped = cgImage.cropping(to: cropRect) else {
      throw RuntimeError.error(withMessage: "Failed to crop image!")
    }
    return UIImage(cgImage: cropped)
  }

  private static func saveToTempFile(_ image: UIImage) throws -> String {
    let url = FileManager.default.temporaryDirectory
      .appendingPathComponent("cropped_\(UUID().uuidString)")
      .appendingPathExtension("jpg")
    guard let data = image.jpegData(compressionQuality: 1.0) else {
      throw RuntimeError.error(withMessage: "Failed to encode image!")
    }
    try data.write(to: url)
    return url.path
  }
}

File: android/src/main/java/com/margelo/nitro/visioncameracropper/VisionCameraCropperFactory.kt
package com.margelo.nitro.visioncameracropper

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.util.Base64
import com.margelo.nitro.NitroModules
import com.margelo.nitro.camera.CameraOrientation
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.public.NativeFrame
import com.margelo.nitro.camera.extensions.toBitmap
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream

@DoNotStrip
@Keep
class VisionCameraCropperFactory : HybridVisionCameraCropperSpec() {

  @DoNotStrip
  @Keep
  override fun rotateImage(
    base64: String,
    degree: Double
  ): String {
    val decode = Base64.decode(base64, Base64.DEFAULT)
    var bitmap = BitmapFactory.decodeByteArray(decode, 0, decode.size)

    if (degree != 0.0) {
      val matrix = Matrix()
      matrix.postRotate(degree.toFloat())

      val newBitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)

      if(newBitmap != bitmap) {
        bitmap.recycle()
      }

      bitmap = newBitmap
    }

    val outputStream = ByteArrayOutputStream()
    bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
    return Base64.encodeToString(outputStream.toByteArray(), Base64.DEFAULT)
  }

  @DoNotStrip
  @Keep
  override fun cropFrame(
    frame: HybridFrameSpec,
    options: CropConfig
  ): CropResult {
    val nativeFrame = (frame as? NativeFrame) ?: throw Exception("Frame is not a NativeFrame!")
    var bitmap = nativeFrame.image.toBitmap(frame.orientation, frame.isMirrored)

    // FOR SOME REASON NEED TO ROTATE WHEN ORIENTATION IS LEFT
    if(frame.orientation == CameraOrientation.LEFT) {
      val matrix = Matrix()
      matrix.postRotate(180.0F)

      val newBitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)

      if(newBitmap != bitmap) {
        bitmap.recycle()
      }

      bitmap = newBitmap
    }

    val cropRegion = options.cropRegion
    if (cropRegion != null) {
      val left = (cropRegion.left / 100.0 * bitmap.width).toInt()
      val top = (cropRegion.top / 100.0 * bitmap.height).toInt()
      val width = (cropRegion.width / 100.0 * bitmap.width).toInt()
      val height = (cropRegion.height / 100.0 * bitmap.height).toInt()
      val croppedBitmap = Bitmap.createBitmap(bitmap, left, top, width, height)

      if (croppedBitmap != bitmap) {
        bitmap.recycle()
      }

      bitmap = croppedBitmap
    }

    var base64: String? = null
    if (options.includeImageBase64 == true) {
      val outputStream = ByteArrayOutputStream()
      bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
      base64 = Base64.encodeToString(outputStream.toByteArray(), Base64.DEFAULT)
    }

    var path: String? = null
    if (options.saveAsFile == true) {
      val context = NitroModules.applicationContext ?: throw Exception("No Context available!")
      val file = File.createTempFile("cropped_", ".jpg", context.cacheDir)
      FileOutputStream(file).use { fos ->
        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fos)
      }
      path = file.absolutePath
    }

    bitmap.recycle()

    return CropResult(base64, path)
  }
}
